### PR TITLE
[Core] Only include undo.h from main.cpp

### PR DESCRIPTION
**Pull request by random-zebra**
Backport bitcoin/bitcoin#5111

Original description:
> Built on top of bitcoin/bitcoin#5100.
> coins.o doesn't need to depend on undo.h (Decouple CCoins from CTxInUndo)
> miner.o neither (Decouple miner.o from CTxUndo)
> Finally, CBlockUndo can be moved from main.h to undo.h if it is first decoupled from CDiskBlockPos.

### CODE REVIEW

**Comment by random-zebra** [comment]
Note: this includes (last commit, move-only) the changes of #1739. So it can either replace it, or it will be rebased after #1739 merge.

**Comment by furszy** [comment]
Closing 1739, liking more your PR.

File: src/txmempool.cpp [hunk 1]
```diff
@@ -680,7 +679,7 @@ void CTxMemPool::check(const CCoinsViewCache* pcoins) const
         } else {
             assert(CheckInputs(entry->GetTx(), state, mempoolDuplicate, false, 0, false, NULL));
             CTxUndo undo;
```
Comment (furszy) [inline src/txmempool.cpp:681]: unused `CTxUndo`

Comment (random-zebra) [reply src/txmempool.cpp]: removed

**Pull request merged**